### PR TITLE
Allow resolving mscorlib with different assembly version numbers

### LINKED ISSUE #24178: Navigate to decompiled sources: 'Unknown result type' in output
**Version Used**: 15.6 Preview 2

**Steps to Reproduce**:

1. Add a reference to **System.Collections.Immutable**
1. Add the following code:

    ```csharp
    var array = ImmutableArray.CreateRange(Enumerable.Range(0, 1), x => x);
    ```

1. Navigate to `ImmutableArray.CreateRange`

**Expected Behavior**:

1. The signature of `CreateRange` is the following:

    ```csharp
    public static ImmutableArray<TResult> CreateRange<TSource, TResult>(ImmutableArray<TSource> items, Func<TSource, TResult> selector)
    ```

2. The decompiled method body does not have comments regarding invalid IL

3. The use of `selector` is proper C# code

**Actual Behavior**:

1. The signature of `CreateRange` is incorrect. Specifically, the generic arguments for the type of `selector` are omitted:

    ```csharp
    public static ImmutableArray<TResult> CreateRange<TSource, TResult>(ImmutableArray<TSource> items, Func selector)
    ```

2. The decompiled method body has comments regarding invalid IL:

    ```csharp
    //IL_0000: Unknown result type (might be due to invalid IL)
    //IL_0029: Unknown result type (might be due to invalid IL)
    //IL_0032: Expected Ref, but got Unknown
    ```

3. When `selector` is used to assign values to `array[i]`, the types and syntax are incorrect (both `(?)` and `(!0)` are incorrect):

    ```csharp
    array[i] = (TResult)((Func)(?)selector).Invoke((!0)items[i]);
    ```

:memo: For reference, here is the complete decompiled method:

```csharp
public static ImmutableArray<TResult> CreateRange<TSource, TResult>(ImmutableArray<TSource> items, Func selector)
{
	//IL_0000: Unknown result type (might be due to invalid IL)
	//IL_0029: Unknown result type (might be due to invalid IL)
	//IL_0032: Expected Ref, but got Unknown
	Requires.NotNull<Func>(selector, "selector");
	int length = items.Length;
	if (length == 0)
	{
		return ImmutableArray.Create<TResult>();
	}
	TResult[] array = new TResult[length];
	for (int i = 0; i < array.Length; i++)
	{
		array[i] = (TResult)((Func)(?)selector).Invoke((!0)items[i]);
	}
	return new ImmutableArray<TResult>(array);
}
```

### PR DESCRIPTION
**Pull request by sharwell**
### Customer scenario

Navigate to Decompiled Sources cannot resolve **mscorlib** when the target of a navigation operation references **netstandard**. This causes cascading failures in the decompilation process leading to unusable output.

### Bugs this fixes

Fixes #24178

### Workarounds, if any

None.

### Risk

Low. The original cause of the problem and dependency on assembly unification is not well understood, but this change has been scoped to only impact **mscorlib**, which ILSpy assumes is always available.

### Performance impact

No performance change except in cases where the new behavior is required for correctness.

### Is this a regression from a previous update?

No.

### Root cause analysis

The behavior was discovered early in this new feature, but remained uncorrected while we attempted to find a stable long-term resolution.

### How was the bug found?

New feature testing.

### Test documentation updated?

A manual test case was added for this scenario.

### CODE REVIEW

**Comment by jasonmalinowski** [comment]
@sharwell I tend to push back on these types of fixes because they tend to mask root causes. Sloppiness we have around other assembly handling is going to require non-trivial refactoring to fix some bugs we have in find references, for example. Do we understand why this was necessary?

(and can we get a test?)

**Comment by sharwell** [comment]
@siegfriedpammer Do you know why version 0.0.0.0 is getting requested here?

**Comment by siegfriedpammer** [comment]
I tried to reproduce this... which version of System.Collections.Immutable did you use? Will look into it...

**Comment by sharwell** [comment]
@siegfriedpammer I was using 1.4.0

**Comment by siegfriedpammer** [comment]
@sharwell Follow-up question: what target framework? netcore2.0, netstandard1.0, netstandard2.0 or portable-net45+win8+wp8+wpa81? Also, it seems the package version 1.4.0 contains an assembly with version 1.2.2, which is kinda weird... is that intended? Thanks!

**Comment by sharwell** [comment]
@siegfriedpammer Here are steps to reproduce

1. Check out sharwell/dotnet-trees@b45e8f3bc057adca7ad7f233fae70e54e51230eb
2. Change the target framework of the test project to **net461** [here](https://github.com/sharwell/dotnet-trees/blob/b45e8f3bc057adca7ad7f233fae70e54e51230eb/TunnelVisionLabs.Collections.Trees.Test/TunnelVisionLabs.Collections.Trees.Test.csproj#L5)
3. On [this line](https://github.com/sharwell/dotnet-trees/blob/b45e8f3bc057adca7ad7f233fae70e54e51230eb/TunnelVisionLabs.Collections.Trees.Test/Immutable/ImmutableTreeListTest.cs#L24), try to navigate to `ImmutableList.Create`

:memo: If you omit step 2 (i.e. use **net452** instead of **net461**), the bug does not occur.

**Comment by siegfriedpammer** [comment]
@sharwell Here are the results of my investigation:

* When changing the target framework, the version of System.Collections.Immutable being decompiled first asks for the netstandard.dll (instead of mscorlib.dll).
* The netstandard.dll that is returned by VS is `%PROGRAMFILES(x86)%\Microsoft Visual Studio\Preview\Community\MSBuild\Microsoft\Microsoft.NET.Build.Extensions\net461\lib\netstandard.dll`.
* In order for us to be able to correctly resolve forwarded types all types used in signatures everywhere, we need to load all transitive references as well. The version of netstandard.dll mentioned above indeed only contains references which have their version set to 0:
grafik

The reason this works in standalone ILSpy is that we manually look for an assembly with the same name but different (= latest) version in the GAC directories if the version is set to 0. See https://github.com/icsharpcode/ILSpy/blob/master/ICSharpCode.Decompiler/DotNetCore/UniversalAssemblyResolver.cs#L134 (note that this code was adapted from Mono.Cecil).

The IAssemblyResolver implementation used in Roslyn replaces all this special behavior. I think this is another instance of the problem: "Please tell me which assemblies would be loaded when executing the code in the assembly being decompiled.", because netstandard.dll is similar to a reference assembly.

File: src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceFileService.cs [hunk 1]
```diff
@@ -218,31 +218,24 @@ public AssemblyDefinition Resolve(AssemblyNameReference name)
...
-                        {
-                            candidateAssemblies.Add(assembly);
-                        }
+                        continue;
+                    }
+
+                    if (assembly.Identity.Version != name.Version
+                        && !string.Equals("mscorlib", assembly.Identity.Name, StringComparison.OrdinalIgnoreCase))
+                    {
+                        // MSBuild treats mscorlib special for the purpose of assembly resolution/unification, where all
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceFileService.cs]: Can you point to the code in GitHub with the same hackery?

**Comment by sharwell** [comment]
It appears the strange behavior is fallout from dotnet/corefx#17363. It is not yet clear why this is expected to work.

**Comment by sharwell** [comment]
@weshaggard @ericstj Can you help here? This new Visual Studio feature is failing to resolve mscorlib version 0.0.0.0. It appears that you had an idea of how this resolution was supposed to work in practice, but I can't find any definition or specification for this behavior so I'm not sure what to do about it.

**Comment by weshaggard** [comment]
You can read more about this issue at https://github.com/dotnet/core-setup/issues/3297. Long story short we reuse the netstandard facade in different contexts where the assembly versions are different so in order to avoid trying to get the version correct we always pick the lowest version which is 0.0.0.0 with the expectation that will unify to whatever version is found in the context it is being used in.

**Comment by sharwell** [comment]
@weshaggard So the rule is, in some cases you are assuming that something downstream from the facade will ignore the 0.0.0.0 version and substitute something else?

1. How can I deterministically detect that an assembly is a facade with fake version numbers in the references?
2. What do I resolve the fake assembly version to?

I need a precise algorithm here because I can't rely on Fusion or assembly binding redirection.

**Comment by weshaggard** [comment]
I don't think you should be detecting this as a facade with fake version number that is just a hack.

The correct resolution is to use the context you have for the library and resolve to the assembly matching the same name in that context. If you don't have any context then you must use some default context to resolve the reference.

This is true for many cases where an application will have assemblies that reference to different versions of a dependency. To resolve them you need to resolve to the version you have in your context  and verify that the one you resolved is of greater than or equal to in version.

**Comment by natidea** [comment]
Approved to merge via https://devdiv.visualstudio.com/DevDiv/_workitems/edit/558385

**Pull request merged**